File: lilac/src/test/java/com/hartwig/hmftools/lilac/misc/LociPositionTest.java
package com.hartwig.hmftools.lilac.misc;

import static com.hartwig.hmftools.common.genome.refgenome.RefGenomeVersion.V37;
import static com.hartwig.hmftools.common.genome.refgenome.RefGenomeVersion.V38;
import static com.hartwig.hmftools.lilac.LilacUtils.calcNucelotideLocus;
import static com.hartwig.hmftools.lilac.ReferenceData.loadHlaTranscripts;
import static com.hartwig.hmftools.lilac.hla.HlaGene.HLA_A;
import static com.hartwig.hmftools.lilac.hla.HlaGene.HLA_B;
import static com.hartwig.hmftools.lilac.hla.HlaGene.HLA_C;

import static org.junit.Assert.assertEquals;

import java.util.List;
import java.util.Map;

import com.hartwig.hmftools.common.gene.TranscriptData;
import com.hartwig.hmftools.lilac.GeneSelector;
import com.hartwig.hmftools.lilac.hla.HlaGene;

import org.junit.Test;

public class LociPositionTest
{
    @Test
    public void testNucleotideLocus()
    {
        Map<HlaGene, TranscriptData> hlaTranscriptMap = loadHlaTranscripts(V37, GeneSelector.MHC_CLASS_1);

        List<TranscriptData> transcripts = hlaTranscriptMap.values().stream().toList();

        TranscriptData aTranscript = hlaTranscriptMap.get(HLA_A);

        int locus = calcNucelotideLocus(transcripts, aTranscript.CodingStart);
        assertEquals(0, locus);

        locus = calcNucelotideLocus(transcripts, aTranscript.CodingEnd);
        assertEquals(1097, locus);

        TranscriptData bTranscript = hlaTranscriptMap.get(HLA_B);

        locus = calcNucelotideLocus(transcripts, bTranscript.CodingEnd);
        assertEquals(0, locus);

        locus = calcNucelotideLocus(transcripts, bTranscript.CodingStart);
        assertEquals(1088, locus);

        TranscriptData cTranscript = hlaTranscriptMap.get(HLA_C);

        locus = calcNucelotideLocus(transcripts, cTranscript.CodingEnd);
        assertEquals(0, locus);

        locus = calcNucelotideLocus(transcripts, cTranscript.CodingStart);
        assertEquals(1100, locus);

        hlaTranscriptMap.clear();

        hlaTranscriptMap = loadHlaTranscripts(V38, GeneSelector.MHC_CLASS_1);

        transcripts = hlaTranscriptMap.values().stream().toList();

        aTranscript = hlaTranscriptMap.get(HLA_A);

        locus = calcNucelotideLocus(transcripts, aTranscript.CodingStart);
        assertEquals(0, locus);

        locus = calcNucelotideLocus(transcripts, aTranscript.CodingEnd);
        assertEquals(1097, locus);

        bTranscript = hlaTranscriptMap.get(HLA_B);

        locus = calcNucelotideLocus(transcripts, bTranscript.CodingEnd);
        assertEquals(0, locus);

        locus = calcNucelotideLocus(transcripts, bTranscript.CodingStart);
        assertEquals(1088, locus);

        cTranscript = hlaTranscriptMap.get(HLA_C);

        locus = calcNucelotideLocus(transcripts, cTranscript.CodingEnd);
        assertEquals(0, locus);

        locus = calcNucelotideLocus(transcripts, cTranscript.CodingStart);
        assertEquals(1100, locus);
    }
}

File: lilac/src/test/kotlin/com/hartwig/hmftools/lilackt/LociPositionTest.kt
package com.hartwig.hmftools.lilackt

import com.hartwig.hmftools.common.genome.genepanel.HmfGenePanelSupplier
import junit.framework.Assert.assertEquals
import junit.framework.Assert.assertTrue
import org.junit.Test

class LociPositionTest {

    private val transcripts = HmfGenePanelSupplier.allGenesMap37()
    private val aTranscript = transcripts[LilacApplication.HLA_A]!!
    private val bTranscript = transcripts[LilacApplication.HLA_B]!!
    private val hlaTranscripts = listOf(aTranscript, bTranscript, transcripts[LilacApplication.HLA_C]!!)
    private val victim = LociPosition(hlaTranscripts)

    @Test
    fun testA() {
        val minLoci = victim.nucelotideLoci(aTranscript.codingStart().toInt())
        assertTrue(minLoci >= 0)

        val maxLoci = victim.nucelotideLoci(aTranscript.codingEnd().toInt())

        for (i in minLoci..maxLoci) {
            val position = victim.position(i, aTranscript)
            val loci = victim.nucelotideLoci(position)
            assertEquals(i, loci)
        }
    }

    @Test
    fun testB() {
        val minLoci = victim.nucelotideLoci(bTranscript.codingEnd().toInt())
        val maxLoci = victim.nucelotideLoci(bTranscript.codingStart().toInt())
        assertTrue(minLoci >= 0)
        assertTrue(maxLoci > minLoci)

        for (i in minLoci..maxLoci) {
            val position = victim.position(i, bTranscript)
            val loci = victim.nucelotideLoci(position)
            assertEquals(i, loci)
        }
    }

}